### PR TITLE
Add ptvsd

### PR DESCRIPTION
**Pull request by dhirschfeld**
- https://github.com/Microsoft/ptvsd
 - https://pypi.org/project/ptvsd/#files

### CODE REVIEW

**Comment by dhirschfeld** [comment]
Hi @zooba,
Are you fine with me adding `ptvsd` to `conda-forge`? Would you like to be added as a maintainer? Being a maintainer would involve a few automated emails from the bot for e.g. merging new update PRs.

**Comment by zooba** [comment]
See whether @mikhailarkhipov or @int19h would rather do it. They're more closely working on ptvsd these days than I am.

**Comment by dhirschfeld** [comment]
ping @MikhailArkhipov / @int19h! I'll go ahead and get this merged shortly as it's easy to add you as maintainers later if you're interested...

**Comment by karthiknadig** [comment]
@dhirschfeld  Please add me and @int19h as maintainers.

**Comment by dhirschfeld** [comment]
Done!

**Comment by karthiknadig** [comment]
@dhirschfeld  can you add me as well?

**Comment by dhirschfeld** [comment]
> *dhirschfeld can you add me as well?*

sorry - copy/paste mistake! Hopefully @MikhailArkhipov doesn't mind being a maintainer also?

**Comment by CJ-Wright** [comment]
@dhirschfeld please ping @conda-forge/staged-recipes when recipes are ready for review

File: recipes/ptvsd/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.zip
+  sha256: cfcde6a3de3cfa720e4f637af13deeae744f6dc6665b9bda92380885caf16ae6
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . --no-deps -vv"
+
+requirements:
```
[the remark this answers is not in the document]
Comment (dhirschfeld) [reply recipes/ptvsd/meta.yaml]: Good catch - thanks! Have added `{{ compiler('c') }}` to the build requirements below.

**Comment by isuruf** [comment]
Why is this package vendoring pydevd?

**Comment by dhirschfeld** [comment]
In case it's of interest there also appears to be a bunch of vendored py2 code which causes byte-compilation to fail. It doesn't seem to be fatal for building the package, but obviously will never work if it's ever imported at runtime.
```
Successfully built ptvsd
Installing collected packages: ptvsd
  *** Error compiling '/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_SimpleXMLRPCServer.py'...
    File "/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_SimpleXMLRPCServer.py", line 263
      except Fault, fault:
                  ^
  SyntaxError: invalid syntax
  
  *** Error compiling '/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_SocketServer.py'...
    File "/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_SocketServer.py", line 339
      print '-'*40
              ^
  SyntaxError: Missing parentheses in call to 'print'. Did you mean print('-'*40)?
  
  *** Error compiling '/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_inspect.py'...
    File "/tmp/pip-install-sajsl0s6/ptvsd/ptvsd/_vendored/pydevd/_pydev_imps/_pydev_inspect.py", line 293
      raise TypeError, 'arg is a built-in module'
                     ^
  SyntaxError: invalid syntax

<snip>
```

**Comment by int19h** [comment]
The intent is for the package to have no external dependencies to minimize the friction when installing it on remote machines possibly with no direct internet connectivity.

With pydevd specifically, we also need to add things to it to support our scenarios, and those then get integrated upstream eventually (but we can't wait for upstream in most cases).

File: recipes/ptvsd/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  run:
+    - python
+
+test:
+  imports:
+    - ptvsd
+
+about:
+  home: https://github.com/Microsoft/ptvsd/
+  license: MIT
```
Comment (isuruf) [inline recipes/ptvsd/meta.yaml]: This is wrong. pydevd is licensed under EPL which makes the combined work EPL in my understanding

Comment (dhirschfeld) [reply recipes/ptvsd/meta.yaml]: I just got that from the [`LICENSE`](https://github.com/Microsoft/ptvsd/blob/master/LICENSE) file. I'm happy to defer to whatever is the correct thing to do here though...

Comment (dhirschfeld) [reply recipes/ptvsd/meta.yaml]: The EPL license is [included in the source](https://github.com/Microsoft/ptvsd/blob/master/src/ptvsd/_vendored/pydevd/LICENSE) - just reading it now...

Comment (int19h) [reply recipes/ptvsd/meta.yaml]: Take a look a this:

https://github.com/Microsoft/ptvsd/blob/master/src/ptvsd/ThirdPartyNotices.txt

Comment (dhirschfeld) [reply recipes/ptvsd/meta.yaml]: [Seems relevant](https://github.com/Microsoft/ptvsd/blob/master/src/ptvsd/_vendored/pydevd/LICENSE#L89-L94):
> When the Program is made available in source code form:
> 
>   a) it must be made available under this Agreement; and
>   b) a copy of this Agreement must be included with each copy of the Program.
>      Contributors may not remove or alter any copyright notices contained within the Program.

**Comment by dhirschfeld** [comment]
@isuruf - are the CI failures a new `conda-build` bug?
```
  File "/opt/conda/lib/python3.6/site-packages/conda_build/os_utils/liefldd.py", line 127, in get_rpaths
    rpaths.append(exe_dirname.replace('\\', '/'))
AttributeError: 'NoneType' object has no attribute 'replace'
```

It seems that the `exe_dirname` being used on [L127](https://github.com/conda/conda-build/blob/master/conda_build/os_utils/liefldd.py#L127) is None:
```python
def get_rpaths(file, exe_dirname, envroot, windows_root=''):
    binary = ensure_binary(file)
    rpaths = []
    if binary:
        if binary.format == lief.EXE_FORMATS.PE:
            # To allow the unix-y rpath code to work we consider
            # exes as having rpaths of env + CONDA_WINDOWS_PATHS
            # and consider DLLs as having no rpaths.
            # .. scratch that, we don't pass exes in as the root
            # entries so we just need rpaths for all files and
            # not to apply them transitively.
            # https://docs.microsoft.com/en-us/windows/desktop/dlls/dynamic-link-library-search-order
rpaths.append(exe_dirname.replace('\\', '/'))
```

Jumping up the stack to where this is called on [L373](https://github.com/conda/conda-build/blob/a842bfe26008661d46c3ca54100244946bc6f4ee/conda_build/os_utils/liefldd.py#L370-L373) in `inspect_linkages_lief`:
```python
if uniqueness_key not in already_seen:
    parent_exe_dirname = None
    if binary.format == lief.EXE_FORMATS.PE:
        tmp_filename = filename2
        while tmp_filename:
            if not parent_exe_dirname and codefile_type(tmp_filename) == 'EXEfile':
                parent_exe_dirname = os.path.dirname(tmp_filename)
            tmp_filename = parents_by_filename[tmp_filename]
    else:
        parent_exe_dirname = exedir
    rpaths_by_binary[filename2] = get_rpaths(binary,
                                             parent_exe_dirname,
                                             envroot.replace(os.sep, '/'),
                                             sysroot)
```

...and I'm a bit mystified what it's actually doing 😕

**Comment by isuruf** [comment]
https://github.com/conda/conda-build/issues/3287#issuecomment-443953991

**Comment by dhirschfeld** [comment]
Given that `pydevd` appears to be licensed separately in [`ThirdPartyNotices.txt`](https://github.com/Microsoft/ptvsd/blob/master/src/ptvsd/ThirdPartyNotices.txt) what should conda-forge do? License as `MIT+EPL`?

**Comment by dhirschfeld** [comment]
@conda-forge-admin, please rerender

**Comment by dhirschfeld** [comment]
@conda-forge-admin, please rerender

**Comment by dhirschfeld** [comment]
@conda-forge-admin, please rerender

**Comment by dhirschfeld** [comment]
@conda-forge-admin please rerender

**Comment by dhirschfeld** [comment]
🤔 does rerendering only work on the actual feedstocks? I guess so...

**Comment by dhirschfeld** [comment]
@conda-forge/core - I think this is ready to go. I've kept the license as MIT as it's the license of the actual code but I've also included `ThirdPartyNotices.txt` as the licenses for the vendored deps.

**Comment by CJ-Wright** [comment]
Please ping @conda-forge/staged-recipes

**Comment by dhirschfeld** [comment]
Sorry about that - I'm always forgetting to ping the right group :grimacing:

**Comment by dhirschfeld** [comment]
> Is the third party licenses file copied into the final conda package?

I'm not sure how to test that with conda-forge as I don't think I have access to the built package? 

When running a normal conda build all files in the recipe directory are copied into the info folder of the built package. Is that not the case for conda forge? How can you tell?

If the file isn't in the built package how can I ensure that it is?

**Comment by beckermr** [comment]
Ahhh ok. I did not know it copied everything by default! LGTM!

**Comment by dhirschfeld** [comment]
Hmmm, the feedstock seems empty :thinking: 
https://github.com/conda-forge/ptvsd-feedstock

Maybe I'm too quick off the mark - will check back shortly...

**Comment by beckermr** [comment]
There has been a conversion problem in staged recipes for a few days now. Usually they resolve themselves but this one seems ongoing.

**Comment by dhirschfeld** [comment]
Just confirming that the `ThirdPartyNotices.txt` file did end up in the built pacakage:

image

Thanks for the merge @beckermr!

**Pull request merged**